### DISCUSSION TITLE
"request.json()" hangs indefinitely in middleware

### DISCUSSION THREAD

**Discussion by leo-aa88** [fastapi/fastapi#8385] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request, Response
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint


class Middleware(BaseHTTPMiddleware):

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        await request.json()
        response = await call_next(request)
        return response


app = FastAPI()
app.add_middleware(Middleware)


@app.post("/test")
async def test(test: dict) -> dict:
    return {"data": "test"}
```

### Description

- Run script with: `uvicorn main:app --reload`
- Run CURL with POST request: `curl -X 'POST'   'http://localhost:8000/test'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{"test": "test"}'`

The application keeps hanging indefinitely and don't send a response to the request. I tested the equivalent code using `app = Starlette()` and it answers correctly.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.79.0

### Python Version

Python 3.10.4

**Comment by virgesmith** [comment]
I have exactly the same problem on fastapi 0.81.0, python 3.10.6 ubuntu 22.04

**Comment by csrgxtu** [comment]
you should remove line 8 code:
```python
await request.json()
```
it is not this code pending your execution, it's following code pending your execution:

```python
@app.post("/test")  # here pending your code
async def test(test: dict) -> dict:
    return {"data": "test"}
```
when fastapi run your test func, it tries to fill the params from request.body, but you already read the body out from stream, thus later read will pending. ref following fastapi code:

![image](https://user-images.githubusercontent.com/5053620/190113608-a09d2a69-3a43-43be-91f4-0f277fad2379.png)

![image](https://user-images.githubusercontent.com/5053620/190114235-009a47ca-4f21-492c-bb92-3a119a2e6bf3.png)

**Comment by virgesmith** [comment]
In my case I actually want to read the json in the middleware (for HMAC validation) before forwarding to the endpoint, is this possible?

**Comment by leo-aa88** [comment]
@csrgxtu I need to read the body in JSON format in several middleware in order to process the data for various reasons (logging, security, sanitizing, etc.) so I wonder what the correct implementation would be in order to avoid halting the program?

**Comment by iudeen** [comment]
In case you want to do this, you need to iterate `re.stream()`. 

I would rather suggest writing a pure ASGI middleware instead.

**Comment by csrgxtu** [comment]
@araujo88 you can ref this issue: https://github.com/tiangolo/fastapi/issues/394

in summary it is a bug or design issue from starlette, and there is work around:

```python
from typing import Mapping

from starlette.requests import Request

from fastapi import FastAPI, APIRouter, Depends

app = FastAPI()
api_router = APIRouter()

@api_router.post("/")
def read_root(arg: Mapping[str, str]):
    return {"Hello": "World"}

async def auth_middleware(request: Request):
    # you can implement your auth here
    print(await request.json())

# the trick here is including auth_middleware in the dependencies:
app.include_router(api_router, dependencies=[Depends(auth_middleware)])
```
thus this middleware logic will work on every requests, we use it same way in our production env, works fine.

note: `Dependencies` will inject your auth logic into each request, for more detail ref: https://fastapi.tiangolo.com/tutorial/dependencies/

**Comment by leo-aa88** [comment]
Is it possible to stream a request multiple times in FastAPI using `self.stream()`? I tried various approches, like copying the `Request` object, but none of them worked.

**Comment by xingdongzhe** [comment]
@araujo88 It can't be using `self.stream()`, for per request, only one time can use `self.stream()`.  
Or you will hang forever because of `await self.event_message.wait()`  
[You can see here: uvicorn/protocols/http/h11_impl.py#L543](https://github.com/encode/uvicorn/blob/fedf67e68d3f81c846bf3740c8ae21fa133000a4/uvicorn/protocols/http/h11_impl.py#L543)  
This is because that when `self.stream()` done, there are no action to notify `self.event_message`
This `asyncio.Event` will be notified of four actions:
- when Connection Lost
- when body_data received
- when body_data readed done
- when write reponse data done

And you can try this:
```
async for chunk in request.stream():
    print(chunk)
await request.json()
```

**Comment by leo-aa88** [comment]
@XingDongZhe I'm trying to process the request body through different middleware for different purposes (such as logging, sanitizing, etc.) however once the `request.stream()` has been called once by the first middleware, the next middleware keeps awaiting for the request due to the way `request.stream()` is implemented. What would be the correct approach for this? I've seen workarounds such as implementing a new middleware class instead of using `BaseHTTPMiddleware` but that seems too far-off to me.

**Comment by iudeen** [comment]
If you want to keep using BaseHTTPMiddleware, I would suggest you create a new Request object at the end of your middleware to avoid such indefinite waits.

**Comment by xingdongzhe** [comment]
@araujo88 Yes, You can workaround with override parameter (or function): `recevie` to cache `request.stream()` data
There is exmple code in issue, but I can't find it now.
You can look for it

**Comment by adriangb** [comment, marked as answer]
This will be solved by #9636

**Comment by wu-clan** [comment]
Hi, @adriangb 

I tracked from https://github.com/encode/starlette/pull/2026 to https://github.com/tiangolo/fastapi/pull/9636

Trying to find a solution, I didn't use BaseHTTPMiddleware due to some circumstances, but tried to write pure ASGI middleware, but unfortunately，app hangs indefinitely

**Reply by adriangb** [reply]
The original issue is about BaseHTTPMiddleware, what you have is a pure ASGI middleware. The cause is the same and the solution is to do something similar to what we did in BaseHTTPMiddleware.

**Reply by adriangb** [reply]
Basically, when you call the downstream app you need to pass in a `receive` that reads off of the requests `body`:

```python
async def wrapped_rcv_gen():
    async for chunk in request.stream():
        yield {“type”: “http.request”, “body”: await request.body()}
        async for msg in request.receive:
            yield msg

wrapped_rcv = wrapped_rcv_gen().__aenext__
```

**Reply by Kludex** [reply]
Add what exactly?

**Reply by wu-clan** [reply]
I think it's here https://www.starlette.io/middleware/#reusing-starlette-components.

Warning everyone about the difference when using request.body() and these things in middleware, because a lot of people are using it, and although BaseHTTPMiddleware is fixed, for people using custom middleware, they always go to github for answers, and, in Starlette's repository, such issue came up n times, and obviously needed some suggested solutions.

I'm not sure if I'm right, but for developers, the documentation will help them a lot.

**Reply by Kludex** [reply]
> such issue came up n times, and obviously needed some suggested solutions.

I don't think this is accurate. I think the issue that came up "n times" was the one with `BaseHTTPMiddleware`. 🤔 

You can create a PR against Starlette if you think it is that obvious. I can't guarantee it will be merged, but I'm willing to review it. 🙏
